README.md
# Lambda-local

[![NPM](https://nodei.co/npm/lambda-local.png?compact=true)](https://www.npmjs.com/package/lambda-local)

[![Lambda-local unit tests](https://github.com/ashiina/lambda-local/actions/workflows/unittests.yml/badge.svg?branch=develop&event=push)](https://github.com/ashiina/lambda-local/actions?query=event%3Apush)

Lambda-local lets you test **NodeJS Amazon Lambda functions** on your local machine, by providing a simplistic API and command-line tool.

It does not aim to be perfectly feature proof as projects like [serverless-offline](https://github.com/dherault/serverless-offline) or [docker-lambda](https://github.com/lambci/docker-lambda), but rather to remain **very light** (it still provides a fully built `Context`, handles all of its parameters and functions, and everything is customizable easily).

The main target are unit tests and running lambda functions locally.

## Install

```bash
npm install -g lambda-local
```

## Build

```bash
make build
```
Or
```bash
npm install
npm install --only=dev
npm run build
```

## Usage

- **As an API:** You can also use Lambda local directly in a script. For instance, it is interesting in a [MochaJS][1] test suite in order to get test coverage.
- **As a command line tool:** You can use Lambda-local as a command line tool.

If you're unsure about some definitions, see [Definitions](#about-definitions) for terminology.

## About: API

### LambdaLocal

API accessible with:
```js
const lambdaLocal = require("lambda-local");
```

Or on TypeScript (supported on 1.7.0+):

```js
import lambdaLocal = require("lambda-local");
```

#### `lambdaLocal.execute(options)`

Executes a lambda given the `options` object, which is a dictionary where the keys may be:

| Key name | Description |
| --- | --- |
| `event`|requested event as a json object|
| `lambdaPath`|requested path to the lambda function|
| `lambdaFunc`|pass the lambda function. You cannot use it at the same time as lambdaPath|
| `profilePath`|optional, path to your AWS credentials file|
| `profileName`|optional, aws profile name. Must be used with |
| `lambdaHandler`|optional handler name, default to `handler`|
| `region`|optional, AWS region, default to `us-east-1`|| `callbackWaitsForEmptyEventLoop`|optional, default to `false`. Setting it to True will wait for an empty loop before returning.|
| `timeoutMs`|optional, timeout, default to 3000 ms|
| `esm`|boolean, marks that the script is an ECMAScript module (use import), default false|
| `environment`|optional, extra environment variables for the lambda|
| `envfile`|optional, load an environment file before booting|
| `envdestroy`|optional, destroy added environment on closing, default to false|
| `verboseLevel`|optional, default 3. Level 2 dismiss handler() text, level 1 dismiss lambda-local text and level 0 dismiss also the result.|
| `callback`|optional, lambda third parameter [callback][1]. When left out a Promise is returned|
| `clientContext`|optional, used to populated clientContext property of lambda second parameter (context)

#### `lambdaLocal.setLogger(logger)`
#### `lambdaLocal.getLogger()`

Those functions allow to access the [winston](https://www.npmjs.com/package/winston) logger used by lambda-local.

## API examples

A lot of examples, especially used among Mocha, may be found in the test files over: [here](https://github.com/ashiina/lambda-local/tree/develop/test)

##### Basic usage: Using Promises

```js
const lambdaLocal = require('lambda-local');

var jsonPayload = {
    'key': 1,
    'another_key': "Some text"
}

lambdaLocal.execute({
    event: jsonPayload,
    lambdaPath: path.join(__dirname, 'path_to_index.js'),
    profilePath: '~/.aws/credentials',
    profileName: 'default',
    timeoutMs: 3000
}).then(function(done) {
    console.log(done);
}).catch(function(err) {
    console.log(err);
});
```

#### Basic usage: using callbacks

```js
const lambdaLocal = require('lambda-local');

var jsonPayload = {
    'key': 1,
    'another_key': "Some text"
}

lambdaLocal.execute({
    event: jsonPayload,
    lambdaPath: path.join(__dirname, 'path_to_index.js'),
    profilePath: '~/.aws/credentials',
    profileName: 'default',
    timeoutMs: 3000,
    callback: function(err, data) {
        if (err) {
            console.log(err);
        } else {
            console.log(data);
        }
    },
    clientContext: JSON.stringify({clientId: 'xxxx'})
});
```

## About: CLI

### Available Arguments
*    `-l, --lambda-path <lambda index path>`            (required) Specify Lambda function file name.
*    `-e, --event-path <event path>`                    (required --watch is not in use) Specify event data file name.
*    `-h, --handler <handler name>`                     (optional) Lambda function handler name. Default is "handler".
*    `-t, --timeout <timeout>`                          (optional) Seconds until lambda function timeout. Default is 3 seconds.
*    `--esm`                                            (optional) Load lambda function as ECMAScript module.
*    `-r, --region <aws region>`                        (optional) Sets the AWS region, defaults to us-east-1.
*    `-P, --profile-path <aws profile name>`            (optional) Read the specified AWS credentials file.
*    `-p, --profile <aws profile name>`                 (optional) Use with **-P**: Read the AWS profile of the file.
*    `-E, --environment <JSON {key:value}>`             (optional) Set extra environment variables for the lambda
*    `--wait-empty-event-loop`                          (optional) Sets callbackWaitsForEmptyEventLoop=True => will wait for an empty loop before returning. This is false by default because our implementation isn\'t perfect and only "emulates" it.
*    `--envdestroy`                                     (optional) Destroy added environment on closing. Defaults to false
*    `-v, --verboselevel <3/2/1/0>`                     (optional) Default 3. Level 2 dismiss handler() text, level 1 dismiss lambda-local text and level 0 dismiss also the result.
*    `--envfile <path/to/env/file>`                     (optional) Set extra environment variables from an env file
*    `--inspect [[host:]port]`                          (optional) Starts lambda-local using the NodeJS inspector (available in nodejs > 8.0.0)
*    `-W, --watch [port]`                               (optional) Starts lambda-local in watch mode listening to the specified port [1-65535].

### CLI examples

```bash
# Simple usage
lambda-local -l index.js -h handler -e examples/s3-put.js

# Input environment variables
lambda-local -l index.js -h handler -e examples/s3-put.js -E '{"key":"value","key2":"value2"}'
```

#### Running lambda functions as a HTTP Server (Amazon API Gateway payload format version 2.0.)

A simple way you can run lambda functions locally, without the need to create any special template files (like Serverless plugin and SAM requires), just adding the parameter `--watch`. It will raise a http server listening to the specified port (default is 8008). You can then call the lambda as mentionned here:
https://docs.aws.amazon.com/lambda/latest/dg/urls-invocation.html

```bash
lambda-local -l examples/handler_gateway2.js -h handler --watch 8008

curl --request POST \
  --url http://localhost:8008/ \
  --header 'content-type: application/json' \
  --data '{
        "key1": "value1",
        "key2": "value2",
        "key3": "value3"
}'
{"message":"This is a response"}
```

You can also use the following format for your event, in order to avoid using the Amazon API Gateway format:
```
{
    "event": {
        "key1": "value1",
        "key2": "value2",
        "key3": "value3"
    }
}
```

In this case, the event will be passed directly to the handler.

## About: Definitions

### Event data
Event sample data are placed in `examples` folder - feel free to use the files in here, or create your own event data.
Event data are just JSON objects exported:

```js
// Sample event data
module.exports = {
    foo: "bar"
};
```

### Context
The `context` object has been sampled from what's visible when running an actual Lambda function on AWS, and the [available documentation](https://docs.aws.amazon.com/lambda/latest/dg/nodejs-prog-model-context.html)
They may change the internals of this object, and Lambda-local does not guarantee that this will always be up-to-date with the actual context object.

### AWS-SDK
As of version 2.0.0, lambda-local no longer packages AWS-SDK in its dependencies. To run a function that makes use of this module, make sure to install it as a dependency in your project.

## Other links

- If you are willing to test an app based on the ASK-SDK, have a look at https://github.com/taimos/ask-sdk-test

## Development

 * Run `make` to install npm modules. (Required to develop & test lambda-local)
 * Run `make test` to execute the mocha test.
* Run `make clean` to reset the repository.

## License

This library is released under the MIT license.

[1]: http://docs.aws.amazon.com/lambda/latest/dg/nodejs-prog-model-handler.html


CHANGELOG.md
# ChangeLog

## 2.1.2 (2023/08/19)
* Fix regression: support for NodeJS>=8
* Fix regression: old events in --watch

## 2.1.1 (2023/07/22)
* Bug fix: ESM path on Windows
* Downgrade commander-js

## 2.1.0 (2023/07/20)
* Support API gateway in --watch
* Support ECMAScript (thanks tdanecker #230)
* Support Lambda streaming (thanks Skn0tt #229)
* Update dependencies

## 2.0.3 (2022/09/12)
* Fix `callbackWaitsForEmptyEventLoop` on recent NodeJS versions (>=16) (#217)
* Support clientContext as an object (#223)
* Remove useless warning (#224)
* Update dependencies

## 2.0.2 (2022/03/08)
* correctly mirror lambda's behaviour when synchronously returning (thanks Snk0tt)
* Update dependencies

## 2.0.1 (2021/12/22)
* Fix `memoryLimitInMB` being wrongly parsed as int
* Update dependencies

## 2.0.0 (2021/06/19)
* **BREAKING CHANGE: removed aws-sdk dependency**. This allows to make `lambda-local` significantly smaller
* Drop Node 8 & Bump depencies
* Minor CLI fixes
* Move CI from Travis to Github

## 1.7.4 (2020/10/12)
* Fix crash when an Error is raised with no stack (#GH202)
* Add --version and `lambdaLocal.version` support
* Improve logging of the --watch option
* Bump dependencies & minor build improvements

## 1.7.3 (2020/06/04)
* Bump dependencies. No API changes

## 1.7.2 (2020/04/13)
* Add --watch parameter
* Bump dependencies

## 1.7.1 (2020/01/28)
* Add support for `_HANDLER`
* Update environment variables computation & consistency with context

## 1.7.0 (2020/01/17)
* Migrate to TypeScript (#191)

## 1.6.3 (2019/07/04)
* Security updates (update dependencies that had vulnerabilities: [mocha](https://github.com/mochajs/mocha/blob/master/CHANGELOG.md#614--2019-04-18)).
* Pick up AWS profile from env `AWS_PROFILE` or `AWS_DEFAULT_PROFILE` (thanks to @illusori)

## 1.6.2 (2019/04/30)
* Documentation changes. No API changes

## 1.6.1 (2019/04/15)
* Update for Winston 3
* Refactor console output manager

## 1.5.3 (2019/04/15)
* Better handling of TimeoutError: now passed to context as other exceptions
* Improved Makefile (thanks to @joechrysler)
* Drop NodeJS < 6, therefore update all dependencies

## 1.5.2 (2018/12/02)
 * New: ability to specify the context manualy (thanks to Nathan Wright)
 * New: Implement proper callbackWaitsForEmptyLoop
 * Fix ARN context environment vars (thanks to @bar-sc)
 * Better outputs management (circular outputs)

## 1.5.1 (2018/06/08)
 * NodeJS >= 10 proper detection
 * Timeout message fix (thanks to @ribizli)

## 1.5.0 (2018/06/06)
 * Make context an object
 * Support multi-threading & nested calls
 * Context is regenerated on each run (e.g. awsRequestId will change)

## 1.4.8 (2018/05/26)
 * Read default AWS config files
 * Improve absolute/relative path finding
 * Handle syntax error in handlers

## 1.4.7 (2018/04/03)
 * Support async functions (thanks to @hoegertn)

## 1.4.6 (2018/03/08)
 * Fix region detection
 * Add --envdestroy option to restore the environment after testing.
 * Support builtin Error object handling
 * More examples & better README

## 1.4.5 (2017/11/15)
 * Better environment variables managing
 * Fix lambda ending rules (context.done/fail/succeed)

## 1.4.4 (2017/10/24)
 * Timeout fix
 * Better context management
 * Add execution time log

## 1.4.3 (2017/09/23)
 * Several token fixes
 * New log system, using a verbose level
 * Bug fixes

## 1.4.2 (2017/03/09)
 * Added promise support
 * Better JSON parse error handling for environment variables

## 1.4.1 (2017/02/09)
 * Fixed critical bug

## 1.4.0 (2017/02/08)
 * Added support for environment variables

## 1.3.0 (2016/11/26)
 * Fixed critical bug (logger)

## 1.2.0 (2016/11/23)
 * Added mock functionality
 * Dropped Node.js v0.1, v0.12 suport

## 1.1.0 (2016/9/15)
 * The default behavior of lambda-local now does not forcefully call the callback function (`-c` option).
 * Added AWS region option `-r`. Defaults to `us-east-1`.
 * Added AWS profile name option `-p`. 

## 1.0.0 (2016/6/10)
 * lambda-local can now be imported as a node module, and be executed from other node.js programs

## 0.0.10 (2016/5/29)
 * Support for Node.js 4.3.2 runtime
 * Added feature to import AWS profile from commandline option
 * Added necessary environment variables 

## 0.0.9 (2016/4/9)
 * Fixed package.json information for npm

## 0.0.8 (2016/4/9)
 * Added support for all properties in the Context object. 
   Supporting everything in here:
   https://docs.aws.amazon.com/en_us/lambda/latest/dg/nodejs-prog-model-context.html

 * Wrote a mocha unit test for the Context object.

## 0.0.7 (2016/2/21)
 * Minor bug fix: Now correctly outputs the error object when Context.done() has an error.

## 0.0.6 (2015/11/13)
 * Added support for `fail` and `succeed` methods

## 0.0.5 (2015/1/21)
 * First release




src/cli.ts
#!/usr/bin/env node

'use strict';

import lambdaLocal = require('./lambdalocal');
import utils = require('./lib/utils');

/*
 * Local executor for Amazon Lambda function
 */
(function() {
    var logger = lambdaLocal.getLogger();

    // process opts
    var program = require('commander');
    program
        .storeOptionsAsProperties()
        .name('lambda-local')
        .version(lambdaLocal.version)
        .option('-l, --lambda-path <lambda index path>', '(required) Lambda function file name.')
        .option('-e, --event-path <path>', '(required if --watch is not in use) Event data file name.')
        .option('-h, --handler <handler name>',
            '(optional) Lambda function handler name. Default is \'handler\'.')
        .option('--esm',
            '(optional) Load lambda function as ECMAScript module.')
        .option('-t, --timeout <timeout seconds>',
            '(optional) Seconds until lambda function timeout. Default is 3 seconds.')
        .option('-r, --region <aws region>',
            '(optional) default set to us-east-1')
        .option('-p, --profile <aws profile name>',
            '(optional) Read the AWS profile to get the credentials from profile name')
        .option('-P, --profile-path <aws credentials path>',
            '(optional) Read the specified AWS credentials file')
        .option('-E, --environment <JSON {key:value}>',
            '(optional) Set extra environment variables for the lambda')
        .option('--wait-empty-event-loop',
            '(optional) Sets callbackWaitsForEmptyEventLoop=True => will wait for an empty loop before ' +
            'returning. This is false by default because our implementation isn\'t perfect and only "emulates" it.')
        .option('--envdestroy',
            '(optional) Destroy added environment on closing. Defaults to false')
        .option('-v, --verboselevel <3/2/1/0>',
            '(optional) Default 3. Level 2 dismiss handler() text, level 1 dismiss lambda-local text ' +
            'and level 0 dismiss also the result.', 3)
        .option('--envfile <path/to/env/file>',
            '(optional) Load additional environment variables from a file')
        .option('--inspect [[host:]port]',
            '(optional) Starts lambda-local using the NodeJS inspector (available in nodejs > 8.0.0)')
        .option('-W, --watch [port]', 
            '(optional) Starts lambda-local in watch mode listening to the specified port [1-65535]. Default is 8008')
        .parse(process.argv);
    
    var eventPath = program.eventPath,
        lambdaPath = program.lambdaPath,
        lambdaHandler = program.handler,
        esm = program.esm,
        profilePath = program.profilePath,
        profileName = program.profile,
        region = program.region,
        environment = program.environment,
        envdestroy = program.envdestroy,
        envfile = program.envfile,
        callbackWaitsForEmptyEventLoop = program.waitEmptyEventLoop,
        verboseLevel = program.verboselevel;

    var port;
    if (program.watch) {
        if (program.watch == true) {
            port = 8008;
        } else {
            port = parseInt(program.watch);
            if(port < 1 || port > 65535) program.help();
        }
        eventPath = true;
    }

    if (!lambdaPath || !eventPath) {
        program.help();
    }

    // default handler name
    if (!lambdaHandler) {
        lambdaHandler = 'handler';
    }
    
    if (environment) {
        try {
            environment = JSON.parse(environment);
        } catch (e) {
            console.log("Invalid environment variable JSON format. ");
            console.log("Example: {\\\"key\\\":\\\"val\\\"\\\,\\\"key2\\\":\\\"val2\\\"}");
            process.exit(1);
        }
    } else {
        environment = [];
    }

    //default callbackWaitsForEmptyEventLoop
    if (!callbackWaitsForEmptyEventLoop) {
        callbackWaitsForEmptyEventLoop = false;
    } else {
        callbackWaitsForEmptyEventLoop = true;
    }

    // timeout milliseconds
    var timeoutMs;
    if (program.timeout) {
        timeoutMs = program.timeout * 1000;
    } else {
        timeoutMs = 3000;
    }

    //Use NodeJS inspector
    var inspector;
    var _close_inspector;
    if (program.inspect) {
        if (utils.get_node_major_version() < 8) {
            logger.log('error', 'Inspector API not available on NodeJS < 8.0.0');
        } else {
            inspector = require('inspector');
            if (program.inspect == true) {
                inspector.open();
            } else if (typeof(program.inspect) === 'string') {
                if (program.inspect.indexOf(":") !== -1){
                    var arr = program.inspect.split(":");
                    inspector.open(parseInt(arr[1]), arr[0]);
                } else {
                    inspector.open(parseInt(program.inspect));
                }
            } else {
                logger.log('error', 'Unknown parameters following --inspect');
            }
            _close_inspector = function(){inspector.close();}
        }
    }

    if (esm) {
        if (utils.get_node_major_version() < 12) {
            console.log("Loading ESCMAScript modules not available on NodeJS < 12.0.0.");
            process.exit(1);
        }
        if (program.watch) {
            console.log("Watch mode not supported for ECMAScript lambda modules.");
            process.exit(1);
        }
    }
    var event = function(){
        if(program.watch) return null;
        return require(utils.getAbsolutePath(eventPath));
    }
    try {
        var init_time = new Date().getTime();
        // execute
        const opts = {
            event: event,
            lambdaPath: lambdaPath,
            lambdaHandler: lambdaHandler,
            esm: esm,
            profilePath: profilePath,
            profileName: profileName,
            region: region,
            callbackWaitsForEmptyEventLoop: callbackWaitsForEmptyEventLoop,
            timeoutMs: timeoutMs,
            environment: environment,
            envdestroy: envdestroy,
            envfile: envfile,
            verboseLevel: verboseLevel,
        };

        if(program.watch) {
            return lambdaLocal.watch({ ...opts, port });
        }

        lambdaLocal.execute({
            ...opts,
            callback: function(err /*, data */) { //data unused
                var exec_time = new Date().getTime() - init_time;
                if (_close_inspector) {
                    _close_inspector();
                }
                if (err !== null && typeof err !== 'undefined') {
                    if (verboseLevel > 0) {
                        logger.log('error', 'Lambda failed in ' + exec_time + 'ms.');
                    }
                    // Finish the process
                    process.exit(1);
                } else {
                    if (verboseLevel > 0) {
                        logger.log('info', 'Lambda successfully executed in ' + exec_time + 'ms.');
                    }
                    process.exit(0);
                }
            }
        });
    } catch (ex) {
        logger.log('error', ex);
        process.exit(1);
    }
})();
